File: carry/lib/src/core/record.dart
import 'clock.dart';

/// Origin of a record or operation.
enum Origin {
  local,
  remote;

  factory Origin.fromJson(String json) {
    return Origin.values.firstWhere((e) => e.name == json.toLowerCase());
  }

  String toJson() => name;
}

/// Metadata attached to every record.
class Metadata {
  Metadata({
    required this.createdAt,
    required this.updatedAt,
    required this.origin,
    required this.clock,
  });

  factory Metadata.fromJson(Map<String, dynamic> json) {
    return Metadata(
      createdAt: json['createdAt'] as int,
      updatedAt: json['updatedAt'] as int,
      origin: Origin.fromJson(json['origin'] as String),
      clock: LogicalClock.fromJson(json['clock'] as Map<String, dynamic>),
    );
  }

  /// When the record was created.
  final int createdAt;

  /// When the record was last updated.
  final int updatedAt;

  /// Origin of the last modification.
  final Origin origin;

  /// Logical clock for ordering.
  final LogicalClock clock;

  Map<String, dynamic> toJson() => {
        'createdAt': createdAt,
        'updatedAt': updatedAt,
        'origin': origin.toJson(),
        'clock': clock.toJson(),
      };
}

/// A data record in the store.
class Record {
  Record({
    required this.id,
    required this.collection,
    required this.version,
    required this.payload,
    required this.metadata,
    required this.deleted,
  });

  factory Record.fromJson(Map<String, dynamic> json) {
    return Record(
      id: json['id'] as String,
      collection: json['collection'] as String,
      version: json['version'] as int,
      payload: json['payload'] as Map<String, dynamic>,
      metadata: Metadata.fromJson(json['metadata'] as Map<String, dynamic>),
      deleted: json['deleted'] as bool,
    );
  }

  /// Unique identifier.
  final String id;

  /// Collection this record belongs to.
  final String collection;

  /// Version number (incremented on each update).
  final int version;

  /// The data payload.
  final Map<String, dynamic> payload;

  /// Record metadata.
  final Metadata metadata;

  /// Whether this record is deleted (tombstone).
  final bool deleted;

  Map<String, dynamic> toJson() => {
        'id': id,
        'collection': collection,
        'version': version,
        'payload': payload,
        'metadata': metadata.toJson(),
        'deleted': deleted,
      };

  /// Whether this record is active (not deleted).
  bool get isActive => !deleted;

  /// Get the creation timestamp.
  int get createdAt => metadata.createdAt;

  /// Get the last update timestamp.
  int get updatedAt => metadata.updatedAt;

  @override
  String toString() =>
      'Record($collection/$id v$version${deleted ? ' [deleted]' : ''})';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Record &&
          runtimeType == other.runtimeType &&
          id == other.id &&
          collection == other.collection &&
          version == other.version;

  @override
  int get hashCode => id.hashCode ^ collection.hashCode ^ version.hashCode;
}

File: engine/src/record.rs
//! Record types for storing data.

use crate::{CollectionName, LogicalClock, RecordId, Timestamp, Version};
use serde::{Deserialize, Serialize};

/// Origin of a record or operation.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Origin {
    /// Created or modified locally
    Local,
    /// Received from remote/server
    Remote,
}

/// Metadata associated with a record.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Metadata {
    /// When the record was first created (milliseconds since epoch)
    pub created_at: Timestamp,
    /// When the record was last updated (milliseconds since epoch)
    pub updated_at: Timestamp,
    /// Whether this record originated locally or from remote
    pub origin: Origin,
    /// Logical clock at the time of last update
    pub clock: LogicalClock,
}

impl Metadata {
    /// Create new metadata for a locally created record.
    pub fn new_local(timestamp: Timestamp, clock: LogicalClock) -> Self {
        Self {
            created_at: timestamp,
            updated_at: timestamp,
            origin: Origin::Local,
            clock,
        }
    }

    /// Create new metadata for a remotely received record.
    pub fn new_remote(timestamp: Timestamp, clock: LogicalClock) -> Self {
        Self {
            created_at: timestamp,
            updated_at: timestamp,
            origin: Origin::Remote,
            clock,
        }
    }

    /// Update metadata for a modification.
    pub fn update(&mut self, timestamp: Timestamp, clock: LogicalClock, origin: Origin) {
        self.updated_at = timestamp;
        self.clock = clock;
        self.origin = origin;
    }
}

/// A data record in the store.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Record {
    /// Unique identifier for this record
    pub id: RecordId,
    /// Collection this record belongs to
    pub collection: CollectionName,
    /// Version number, incremented on each update
    pub version: Version,
    /// The actual data payload (JSON value)
    pub payload: serde_json::Value,
    /// Record metadata
    pub metadata: Metadata,
    /// Soft delete flag (tombstone)
    pub deleted: bool,
}

impl Record {
    /// Create a new record.
    pub fn new(
        id: impl Into<RecordId>,
        collection: impl Into<CollectionName>,
        payload: serde_json::Value,
        timestamp: Timestamp,
        clock: LogicalClock,
    ) -> Self {
        Self {
            id: id.into(),
            collection: collection.into(),
            version: 1,
            payload,
            metadata: Metadata::new_local(timestamp, clock),
            deleted: false,
        }
    }

    /// Check if record is active (not deleted).
    pub fn is_active(&self) -> bool {
        !self.deleted
    }

    /// Mark record as deleted (tombstone).
    pub fn mark_deleted(&mut self, timestamp: Timestamp, clock: LogicalClock, origin: Origin) {
        self.deleted = true;
        self.version += 1;
        self.metadata.update(timestamp, clock, origin);
    }

    /// Update record payload.
    pub fn update_payload(
        &mut self,
        payload: serde_json::Value,
        timestamp: Timestamp,
        clock: LogicalClock,
        origin: Origin,
    ) {
        self.payload = payload;
        self.version += 1;
        self.metadata.update(timestamp, clock, origin);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn create_record() {
        let clock = LogicalClock::with_counter("node-1", 1);
        let record = Record::new("user-1", "users", json!({"name": "Alice"}), 1000, clock);

        assert_eq!(record.id, "user-1");
        assert_eq!(record.collection, "users");
        assert_eq!(record.version, 1);
        assert_eq!(record.payload, json!({"name": "Alice"}));
        assert!(!record.deleted);
        assert!(record.is_active());
    }

    #[test]
    fn update_record() {
        let clock = LogicalClock::with_counter("node-1", 1);
        let mut record = Record::new("user-1", "users", json!({"name": "Alice"}), 1000, clock);

        let new_clock = LogicalClock::with_counter("node-1", 2);
        record.update_payload(
            json!({"name": "Alice Smith"}),
            2000,
            new_clock,
            Origin::Local,
        );

        assert_eq!(record.version, 2);
        assert_eq!(record.payload, json!({"name": "Alice Smith"}));
        assert_eq!(record.metadata.updated_at, 2000);
    }

    #[test]
    fn delete_record() {
        let clock = LogicalClock::with_counter("node-1", 1);
        let mut record = Record::new("user-1", "users", json!({"name": "Alice"}), 1000, clock);

        let delete_clock = LogicalClock::with_counter("node-1", 2);
        record.mark_deleted(2000, delete_clock, Origin::Local);

        assert!(record.deleted);
        assert!(!record.is_active());
        assert_eq!(record.version, 2);
    }

    #[test]
    fn metadata_origin() {
        let clock = LogicalClock::with_counter("node-1", 1);

        let local = Metadata::new_local(1000, clock.clone());
        assert_eq!(local.origin, Origin::Local);

        let remote = Metadata::new_remote(1000, clock);
        assert_eq!(remote.origin, Origin::Remote);
    }

    #[test]
    fn serialization_roundtrip() {
        let clock = LogicalClock::with_counter("node-1", 1);
        let record = Record::new(
            "user-1",
            "users",
            json!({"name": "Alice", "age": 30}),
            1000,
            clock,
        );

        let json = serde_json::to_string(&record).unwrap();
        let parsed: Record = serde_json::from_str(&json).unwrap();

        assert_eq!(record, parsed);
    }
}
